### ISSUE TITLE
~!~ OpenSpiel Extravaganza ~!~  Jun 22nd - Jul 3rd

### ISSUE THREAD

**Issue by lanctot** [deepmind/open_spiel#251]
Hello everyone, 

Starting today until July 3rd, there will be concentrated effort to add functionality to OpenSpiel. During this time, we will be updating github every day (possibly even many times per day).

We will be more available than usual to answer questions that you may have about OpenSpiel. Please use this thread.

Among the list of planned additions:

- Public states sub-API (for implementing algorithms such as DeepStack)
- Card games and RL (Hearts, Skat, Bridge, Tarok)
- More games! Clobber, Emergent Communication, Emerald Mines / Boulderdash, Hyper-backgammon)
- Add bidding to Skat
- Get C++ AlphaZero working externally
- AlphaZero in Hex; Hex & AlphaZero improvements
- Initial investigation in convergence of CFR in n-player and general-sum games

We will also try to run some algorithms and might report a few results.

**Comment by lanctot** [comment]
@findmyway did you know about AlphaZero.jl project? I posted a suggestion to support OpenSpiel games: https://github.com/jonathan-laurent/AlphaZero.jl/issues/15 . Would that be something you'd be interested in trying out or looking into, even if you don't have time between this specific period? Would be a neat collaboration between projects!

**Comment by findmyway** [comment]
Hi @lanctot , 
Yeah, in fact, I contacted @jonathan-laurent just after I **decided** to write this Julia wrapper. And we had some nice discussions several months ago https://github.com/jonathan-laurent/AlphaZero.jl/issues/4 . I'll provide more info in the issue you created there. https://github.com/jonathan-laurent/AlphaZero.jl/issues/15

**Comment by lanctot** [comment]
Ok so we will use this thread as a way to describe our project in a bit more detail and show any relevant progress. 

I will start.  I implemented HyperBackgammon, which is a simple variant where each side has only 3 checkers. However, it is not the full game because we still don't have the doubling cube implemented (note to self: we should do that :-p)

Here's a screenshot of the initial state:
```
+------|------+
|......|...ooo|
|......|......|
|......|......|
|......|......|
|......|......|
|      |      |
|......|......|
|......|......|
|......|......|
|......|......|
|......|...xxx|
+------|------+
Turn: *
Dice: 
Bar:
Scores, X: 0, O: 0
```

It is implemented as a variant of Backgammon, you can load it using the game string `backgammon(hyper_backgammon=true)`

**Comment by lanctot** [comment]
With the help of several people @finbarrtimbers @jblespiau , I did the final stages of getting OpenSpiel fully supported with TF 2.2 and Python 3.8. As a result, we now fully support Ubuntu 20.04. For details see https://github.com/deepmind/open_spiel/issues/166 and https://github.com/deepmind/open_spiel/pull/249.

![image](https://user-images.githubusercontent.com/3929263/86146717-d2667b80-bab5-11ea-8ac7-cbac26d50260.png)

**Comment by inejc** [comment]
I finalized de/serialization of CFR and MCCFR C++ solvers and exposed the functionality in Python. For details see https://github.com/deepmind/open_spiel/pull/222.

Below is the output of the Python example where an MCCFR solver is pickled halfway through training and then the process is continued with the loaded solver:
```bash
$ python open_spiel/python/examples/mccfr_cpp_example.py --sampling=external --iterations=20
Iteration 0 exploitability: 0.583333
Iteration 1 exploitability: 0.604167
Iteration 2 exploitability: 0.611111
Iteration 3 exploitability: 0.562500
Iteration 4 exploitability: 0.529167
Iteration 5 exploitability: 0.465278
Iteration 6 exploitability: 0.443452
Iteration 7 exploitability: 0.432292
Iteration 8 exploitability: 0.425926
Iteration 9 exploitability: 0.420833
Persisting the model...
Loading the model...
Exploitability of the loaded model: 0.420833
Iteration 10 exploitability: 0.416667
Iteration 11 exploitability: 0.411458
Iteration 12 exploitability: 0.400819
Iteration 13 exploitability: 0.378803
Iteration 14 exploitability: 0.359722
Iteration 15 exploitability: 0.343027
Iteration 16 exploitability: 0.330226
Iteration 17 exploitability: 0.318848
Iteration 18 exploitability: 0.305478
Iteration 19 exploitability: 0.291289
```

**Comment by inejc** [comment]
@timsmole and I implemented the [Slovenian Tarok card game](https://github.com/semanticweights/tarok) and started porting it to the OpenSpiel repository. For details see https://github.com/deepmind/open_spiel/pull/274.

Below is part of the initial output of the Python script that enables one to play the full game:
```bash
$ python tarok/python/play_game.py                                          
Game phase: GamePhase.CARD_DEALING
Selected contract: Contract.NOT_SELECTED
Current player: -1

Player cards: []

Legal actions: [('Deal', 0)]

Enter action: 0
---------------------------------------------------------------------- 

Game phase: GamePhase.BIDDING
Selected contract: Contract.NOT_SELECTED
Current player: 1

Player cards: [('V', 4), ('VI', 5), ('VII', 6), ('XII', 11), ('XIV', 13), ...]

Legal actions: [('Pass', 0), ('Two', 3), ('One', 4), ('Beggar', 8), ...]

Enter action: 0
---------------------------------------------------------------------- 

Game phase: GamePhase.BIDDING
Selected contract: Contract.NOT_SELECTED
Current player: 2

Player cards: [('Pagat', 0), ('II', 1), ('IIII', 3), ('VIII', 7), ...]

Legal actions: [('Pass', 0), ('Two', 3), ('One', 4), ('Beggar', 8), ...]

Enter action: 0
---------------------------------------------------------------------- 

Game phase: GamePhase.BIDDING
Selected contract: Contract.NOT_SELECTED
Current player: 0

Player cards: [('III', 2), ('X', 9), ('XIII', 12), ('XV', 14), ...]

Legal actions: [('Klop', 1), ('Three', 2), ('Two', 3), ('One', 4), ...]

Enter action: 1
---------------------------------------------------------------------- 

Game phase: GamePhase.TRICKS_PLAYING
Selected contract: Contract.KLOP
Current player: 0

Player cards: [('III', 2), ('X', 9), ('XIII', 12), ('XV', 14), ...]

Trick cards: []

Legal actions: [('III', 2), ('X', 9), ('XIII', 12), ('XV', 14), ...]

Enter action: 2
---------------------------------------------------------------------- 

Game phase: GamePhase.TRICKS_PLAYING
Selected contract: Contract.KLOP
Current player: 1

Player cards: [('IIII', 3), ('V', 4), ('IX', 8), ('XVII', 16), ...]

Trick cards: [('III', 2)]

Legal actions: [('IIII', 3), ('V', 4), ('IX', 8), ('XVII', 16), ...]
```

**Comment by ryanbhayward** [comment]
I've been following what Bedir and Mohammad-R have been doing.  (It's not yet directly related to openspiel, but I've been writing some postscript so that I can generate board position pix for the game of Y, especially on the geodesic-dome-style boards introduced by Craige Schensted)
![Screen Shot 2020-07-02 at 12 19 52 PM](https://user-images.githubusercontent.com/3485282/86395971-5e1b0c00-bc5e-11ea-9e67-93d78d8cbb48.png)
![Screen Shot 2020-07-02 at 12 20 56 PM](https://user-images.githubusercontent.com/3485282/86396060-81de5200-bc5e-11ea-9c0b-ca99bed864a4.png)

**Comment by christianjans** [comment]
I just have added the standard implementation of the game [Clobber](https://en.wikipedia.org/wiki/Clobber).

It supports up to a 99x26 board (99 rows, 26 columns), although boards for humans and computers are usually kept to 5x6, 8x8, and 10x10.

For this original version, beginning from a custom board is not supported in Python, and the option to invert the default checkerboard pattern of the pieces is also not implemented. However, these shouldn't be too difficult to get going and can be done soon.

Below is a full game of me (Player 1, 'x') losing to an MCTS player (Player 0, 'o') on a 4x4 board.

```
$ python3 open_spiel/python/examples/mcts.py --game=clobber\(rows=4,columns=4\) --player1=mcts --player2=human
Initial state:
4xoxo
3oxox
2xoxo
1oxox
 abcd

Player 0 sampled action: c1b1
Next state:
4xoxo
3oxox
2xoxo
1oo.x
 abcd

Choose an action (empty to print legal actions): a4a3
Player 1 sampled action: a4a3
Next state:
4.oxo
3xxox
2xoxo
1oo.x
 abcd

Player 0 sampled action: d2d3
Next state:
4.oxo
3xxoo
2xox.
1oo.x
 abcd

Choose an action (empty to print legal actions): c4b4
Player 1 sampled action: c4b4
Next state:
4.x.o
3xxoo
2xox.
1oo.x
 abcd

Player 0 sampled action: b2c2
Next state:
4.x.o
3xxoo
2x.o.
1oo.x
 abcd

Choose an action (empty to print legal actions): b3c3
Player 1 sampled action: b3c3
Next state:
4.x.o
3x.xo
2x.o.
1oo.x
 abcd

Player 0 sampled action: c2c3
Next state:
4.x.o
3x.oo
2x...
1oo.x
 abcd

Choose an action (empty to print legal actions): a2a1
Player 1 sampled action: a2a1
Next state:
4.x.o
3x.oo
2....
1xo.x
 abcd

Player 0 sampled action: b1a1
Next state:
4.x.o
3x.oo
2....
1o..x
 abcd

Returns: 1.0 -1.0 , Game actions: c1b1 a4a3 d2d3 c4b4 b2c2 b3c3 c2c3 a2a1 b1a1
Number of games played: 1
Number of distinct games played: 1
Players: mcts human
Overall wins [1, 0]
Overall returns [1.0, -1.0]
```

**Comment by tuero** [comment]
I implemented a bridged version of Boulder Dash/Emerald Mines (see [here ](https://www.c64-wiki.com/wiki/Boulder_Dash)for a description, and [here ](https://www.youtube.com/watch?v=FiEVfa1OK_o)for a playthrough). The mechanics of the game are simple: collect enough diamonds to open the exit, while avoiding enemies. This game offers many interesting challenges for AI
- Long delays between rewards
- Dodging enemies by exploiting their predictive movement
- Dropping boulders onto enemies to reveal diamonds (understanding mechanics, consequences of actions, and timing)
- While the agent moves through the level, dirt is removed which causes items to fall, which can result in deadlocks (blocking the exit or diamonds)
- It is also quite fun to play :)

Many of the original elements are supported, and custom boards can be taken in as arguments. I have also converted many of the original Boulder Dash levels into the specified format, which you can find [here ](https://github.com/tuero/bd_open_spiel_levels)(althrough these will be quite hard on their own for AI to tackle). I will also be adding simplified versions of the levels to that repo, which tackle different complexities the game offers while being _easier_ to solve.

I have attached a gif below of the ASCII board. 
![bd_mines_small](https://user-images.githubusercontent.com/12579920/86410675-342b0f00-bc89-11ea-8bf7-a97b7f9fc244.gif)

**Comment by vbhatt-cs** [comment]
I implemented a modified version of the Lewis Signaling Game ([https://en.wikipedia.org/wiki/Lewis_signaling_game](https://en.wikipedia.org/wiki/Lewis_signaling_game)) with an option for arbitrary payoffs. This game can be used as a starting point for multi-agent communication algorithms.

With certain payoff matrices, even a game with 3 states and 3 messages can be challenging for decentralized algorithms. Below are the results for centralized and decentralized tabular Q-Learning and decentralized DQN on identity payoff matrices. Since ε is decayed linearly from 1 to 0, the curve looks nearly linear in the centralized case.

<img src="https://user-images.githubusercontent.com/12762312/86425480-8af11280-bca2-11ea-9cf9-0a58366ca7cd.png" width="400"> <img src="https://user-images.githubusercontent.com/12762312/86425510-9e9c7900-bca2-11ea-8b18-7705b52d8e0a.png" width="400">

<img src="https://user-images.githubusercontent.com/12762312/86425505-9ba18880-bca2-11ea-805a-50ba396c88bb.png" width="400"> <img src="https://user-images.githubusercontent.com/12762312/86425512-9f350f80-bca2-11ea-8c8f-9571241fb298.png" width="400">

The following figures show the final joint policy for each state. The counts for each (state, action) pair denotes the number of runs that had that particular joint policy.

<img src="https://user-images.githubusercontent.com/12762312/86425998-d5bf5a00-bca3-11ea-9d13-61210f82fb25.png" width="250"> <img src="https://user-images.githubusercontent.com/12762312/86425997-d526c380-bca3-11ea-97a2-c3c0010f023b.png" width="250"> <img src="https://user-images.githubusercontent.com/12762312/86426000-d5bf5a00-bca3-11ea-823a-0583ca5b2991.png" width="250">

With the following payoff matrix, the decentralized methods fail to reach the optimal policy. 

<img src="https://user-images.githubusercontent.com/12762312/86426144-2d5dc580-bca4-11ea-8f0a-05a6deddc227.png" width="200">

<img src="https://user-images.githubusercontent.com/12762312/86426222-567e5600-bca4-11ea-8644-d9f1ce603b23.png" width="400"> <img src="https://user-images.githubusercontent.com/12762312/86426223-5716ec80-bca4-11ea-9703-1068a4a6866d.png" width="400">

<img src="https://user-images.githubusercontent.com/12762312/86426221-567e5600-bca4-11ea-83e5-b4d98a41c47f.png" width="400"> <img src="https://user-images.githubusercontent.com/12762312/86426216-55e5bf80-bca4-11ea-9a7c-1067c7bed464.png" width="400">

<img src="https://user-images.githubusercontent.com/12762312/86426218-567e5600-bca4-11ea-9620-1e9a55b01b3c.png" width="250"> <img src="https://user-images.githubusercontent.com/12762312/86426217-567e5600-bca4-11ea-828a-c6f8fe5d75b5.png" width="250"> <img src="https://user-images.githubusercontent.com/12762312/86426215-55e5bf80-bca4-11ea-9562-b41a32831522.png" width="250">

**Comment by jhtschultz** [comment]
As part of the Hearts project, we implemented the classic card game Hearts in OpenSpiel, open sourced the current state of the art Hearts program xinxin (https://github.com/nathansttt), and built an interface that allows xinxin to play as an OpenSpiel bot.

Example OpenSpiel Hearts state:
<img width="167" alt="Screen Shot 2020-07-02 at 10 25 24 PM" src="https://user-images.githubusercontent.com/39270229/86425365-f511b380-bcb2-11ea-81aa-ec4e9afeaf58.png">
Xinxin screenshot:
<img width="375" alt="xinxin copy" src="https://user-images.githubusercontent.com/39270229/86424893-bb8c7880-bcb1-11ea-84c4-428e78f0e1cc.png">

We also added information state resampling to Hearts, which enables algorithms like IS-MCTS to run on the game. Along with xinxin, this provides another useful baseline to compare learning algorithms against. Beyond establishing the first open source benchmarks for Hearts, OpenSpiel-xinxin enables us to generate high-quality games which can be used to learn supervised policies. We wrote a python example of a script that does this. Going forward, we could use these policies to seed learning algorithms, or as a faster alternative to xinxin for move generation in MCTS rollouts.

Hearts is an important representative of the class of trick taking card games. Unlike Bridge, there are no teams in Hearts, making it a multiplayer general sum game. As such, it inherits many theoretical complexities and ambiguities. Although RL research on hearts dates back to at least 1997, it still remains a domain in which algorithms have yet to achieve superhuman performance. With this project we aim to preserve an important part of AI history, and use it to facilitate progress in the active research area of multi-agent RL.

**Comment by ElnazDavoodi** [comment]
We worked on Public states sub-API to add factored-observation support. In imperfect information games, each agent gets different private information, as well as some public information. The agent's observation is factored to "private observation" and "public observation". During an imperfect information game, each agent should have a non-empty "private observation" at some states (e.g. in Kuhn-poker, agents are dealt cards and the card of each agent is its private observation), and both agents have "public observation" in all states of the game (e.g. in Kuhn-poker, both agents can observe all agent's actions such as bet, call, etc.).
Also, a version of Kuhn-poker was added which supports factored-observations.

**Comment by lanctot** [comment]
Hi all, 

So other than importing PRs and running stuff, I managed to get a small but fun thing done: progress on CFR's empirical convergence outside two-player zero-sum games. 

I had previously implemented the `EFCCEDist` and `EFCEDist` functions in `algorithms/corr_dist.*` to be analogues of NashConv: distances to extensive-form coarse-correlated equilibria (EFCCE) and to extensive-form correlated equilibria (EFCE). Motivated by my enthusiasm after reading [Celli et al. 2019](https://arxiv.org/abs/1910.06228) and [Farina et al. '2019](https://arxiv.org/abs/1908.09893), I wanted to know what CFR's emprical joint policy converges to.

So, during the extravaganza, I added the following (to be submitted in an update in the coming weeks): the ability to extract CFR's current joint policy and `corr_dev_builder`, a helper class that builds up correlation devices (distributions over joint pure policies) incrementally over time, with the ability to sample pure joint policies from CFR's stochastic policy. 

I ran CFR on the general-sum variant of 3-card Goofspiel described in Farina et al. above, the specific game string being `turn_based_simultaneous_game(game=goofspiel(num_cards=3,points_order=descending,returns_type=total_points))` . This graph shows the EFCCEDist and EFCEDist of CFR's joint average policy (to build this I sampled 100 joint policies per CFR's mixed strategy and weighed them all equally):

![image](https://user-images.githubusercontent.com/3929263/86436654-ee3d6d80-bcbf-11ea-86b8-1ead8bf2e15a.png)

And this graph shows the expect payoffs to each player:

![image](https://user-images.githubusercontent.com/3929263/86436705-0614f180-bcc0-11ea-8d9d-a50b734c44d2.png)

**Comment by BedirT** [comment]
Hi all,

I took the first steps on modified AlphaZero. I added the first version for alpha-beta search functionality instead of MCTS on AlphaZero. The main idea is to be able to compare the results and see the differences between the search methods. The policy is not used in the search yet (to be added), for now only values are used. It is significantly slower, and depth limited, therefore I couldn't get the results of the experiments yet, which I will be posting when I hopefully do.

**Comment by michalsustr** [comment]
I worked with Elnaz on adding support for factored observations, which are the basis of the [Factored Observation Games (FOG) formalism](https://arxiv.org/abs/1906.11110). We implemented factored observations for Kuhn Poker with arbitrary number of players. We added a special `PublicState` interface and implemented it for Kuhn Poker as well. See [Public State API](https://github.com/deepmind/open_spiel/tree/master/open_spiel/public_states) for more details. This interface provides data structures and methods necessary for algorithms that solve large-scale games of imperfect information, like Heads-Up, No-Limit Texas Hold'em Poker (a game with ~10^160 states). We added python bindings that support efficient vectorized operations via numpy (resp. Eigen) arrays without copy overhead. This allows to create relatively efficient algorithms also in Python, expanding the accessibility of this research area. 

We implemented a `PublicState` <-> `State` visualization that nicely shows the relationship between the game structures:

![Kuhn Poker for 2 players](https://user-images.githubusercontent.com/1756494/86442605-49289200-bccb-11ea-9e21-cfa3373b418d.png)

We have a work-in-progress implementation of the CFR algorithm that runs on the public tree. We expect to publish it within a few days, maybe even tomorrow. We also added a number of consistency tests between the `PublicState` and `State` API.

There remains much to do, but this was a nice first step. Hopefully it will be followed by contributions of many authors, and many scientific papers! :)

**Comment by mrdaliri** [comment]
Hi,
During Extravaganza, I worked on getting C++ AlphaZero to work externally. The current version of AlphaZero is implemented using Tensorflow, which is known to have a tricky C/C++ API.

After many trials and errors, and with the help of many great people, I was able to fix all compile and linking errors for all four TF/C++ targets of OpenSpiel (`alpha_zero` library, `vpnet_test`, `alpha_zero_example` and `tf_trajectories_example`) and build them using Tensorflow 2.2.0. Since TF doesn't have a native C++ API (it only has C API), we utilize [tensorflow_cc](https://github.com/FloopCZ/tensorflow_cc) project to build TF using CMake and Bazel from scratch.

Right now, we still have two runtime errors reported by TF checks. I guess that they are caused by some version mismatch between external libraries (TF/Eigen) and can be fixed by spending more time digging into TF/Eigen source code and release notes.

I'll continue to work on these issues, and I hope we can have a working C++ AlphaZero in a couple of weeks. You can follow updates on this project in issue #172.

**Comment by flostim** [comment]
Hi,

I spent most of the Extravaganza adding a GUI for Skat and Hearts in the soon to be released OpenSpiel GUI framework. They allow you to play the games against bots and see all the observations that a bot would see. Additionally they can be used to review games played by bots.

**Skat:**
![skat-screenshot](https://user-images.githubusercontent.com/11563594/86485057-a23a0980-bd4f-11ea-9c8d-f21bac5327a4.png)

**Hearts:**
![hearts-screenshot](https://user-images.githubusercontent.com/11563594/86485035-93ebed80-bd4f-11ea-8e0e-57049ab3af42.png)

I also started working on adding full bidding to Skat. This turned out to be a bit more complicated than I thought and I will make the game configurable to allow both simplified and full bidding.
